File: frontend-old-egui/src/format.rs
//! Per-column value formatters.
//!
//! A column's metadata may carry a `formatter` JSON blob selecting a typed transform
//! that reshapes each cell's display text. Cell values arrive as plain strings (Arrow's
//! `ArrayFormatter` output), so every formatter parses the string it is given and
//! returns `None` when it can't — the caller then falls back to the raw value.

use jiff::civil::{Date, DateTime};
use jiff::tz::TimeZone;
use jiff::{Timestamp, Unit, Zoned};
use serde::Deserialize;

/// A typed value formatter, parsed from a column's `formatter` metadata blob. The
/// `type` discriminant selects the variant; each variant carries its own config.
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub(crate) enum Formatter {
    /// Formats a number to a min/max number of fraction digits (no thousands grouping).
    #[serde(rename_all = "camelCase")]
    Number {
        #[serde(default)]
        decimal_places: DecimalPlaces,
    },
    /// Formats a timestamp via a jiff `strftime` format string.
    Timestamp { format: String },
    /// Formats a number of seconds as `M:SS`.
    Duration {},
    /// Formats a timestamp as the time elapsed relative to now.
    #[serde(rename_all = "camelCase")]
    RelativeTime {
        #[serde(default)]
        units: Vec<RelativeUnit>,
    },
}

/// The user-facing `decimalPlaces` spec: a plain number, or a one/two-element list of
/// `[min]` / `[min, max]` fraction digits.
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(untagged)]
pub(crate) enum DecimalPlaces {
    Single(usize),
    List(Vec<usize>),
}

impl Default for DecimalPlaces {
    fn default() -> Self {
        DecimalPlaces::List(Vec::new())
    }
}

/// A unit available to the `relativeTime` formatter.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "lowercase")]
pub(crate) enum RelativeUnit {
    Minutes,
    Hours,
    Days,
    Weeks,
    Months,
    Years,
}

impl Formatter {
    /// Formats `value` for display, or returns `None` when it can't be parsed (the
    /// caller then renders the raw value unchanged).
    pub(crate) fn format(&self, value: &str) -> Option<String> {
        match self {
            Formatter::Number { decimal_places } => {
                let (min, max) = resolve_decimals(decimal_places);
                format_number(value, min, max)
            }
            Formatter::Timestamp { format } => format_timestamp(value, format),
            Formatter::Duration {} => format_duration(value),
            Formatter::RelativeTime { units } => {
                let then = parse_zoned(value)?;
                relative_between(&then, &Zoned::now(), units)
            }
        }
    }
}

/// Resolves a [`DecimalPlaces`] spec into a `(min, max)` pair, swapping if min > max.
fn resolve_decimals(spec: &DecimalPlaces) -> (usize, usize) {
    let (mut min, mut max) = match spec {
        DecimalPlaces::Single(n) => (*n, *n),
        DecimalPlaces::List(list) => match list.as_slice() {
            [] => (0, 0),
            [n] => (*n, *n),
            [min, max, ..] => (*min, *max),
        },
    };
    if min > max {
        std::mem::swap(&mut min, &mut max);
    }
    (min, max)
}

/// Formats a number to at most `max` and at least `min` fraction digits, trimming
/// trailing zeros in between. No thousands grouping.
fn format_number(value: &str, min: usize, max: usize) -> Option<String> {
    let n: f64 = value.trim().parse().ok()?;
    let mut s = format!("{n:.max$}");
    // `s` now has exactly `max` fraction digits. Trim trailing zeros down to `min`.
    if max > min
        && let Some(dot) = s.find('.')
    {
        // Shortest acceptable length: the integer part, plus a dot and `min` digits.
        let min_len = if min == 0 { dot } else { dot + 1 + min };
        while s.len() > min_len && s.ends_with('0') {
            s.pop();
        }
        if s.ends_with('.') {
            s.pop();
        }
    }
    Some(s)
}

/// Formats a number of seconds as `M:SS` (minutes unpadded, seconds zero-padded),
/// rounded to the nearest second.
fn format_duration(value: &str) -> Option<String> {
    let secs: f64 = value.trim().parse().ok()?;
    let total = secs.round() as i64;
    let sign = if total < 0 { "-" } else { "" };
    let total = total.abs();
    let minutes = total / 60;
    let seconds = total % 60;
    Some(format!("{sign}{minutes}:{seconds:02}"))
}

/// Formats a parsed timestamp via a jiff `strftime` format string. Tries a civil
/// datetime, then a date, then integer epoch seconds.
fn format_timestamp(value: &str, format: &str) -> Option<String> {
    let v = value.trim();
    // Arrow renders datetimes space-separated (`2026-06-07 14:30:45`); ISO uses `T`.
    if let Ok(dt) = v.replacen(' ', "T", 1).parse::<DateTime>() {
        return Some(dt.strftime(format).to_string());
    }
    if let Ok(date) = v.parse::<Date>() {
        return Some(date.strftime(format).to_string());
    }
    let secs: i64 = v.parse().ok()?;
    let zoned = Timestamp::from_second(secs)
        .ok()?
        .to_zoned(TimeZone::system());
    Some(zoned.strftime(format).to_string())
}

/// Parses a cell value into a zoned instant, for relative-time math. Tries a civil
/// datetime, then a date (start of day), then integer epoch seconds — all in the
/// system time zone.
fn parse_zoned(value: &str) -> Option<Zoned> {
    let v = value.trim();
    if let Ok(dt) = v.replacen(' ', "T", 1).parse::<DateTime>() {
        return dt.to_zoned(TimeZone::system()).ok();
    }
    if let Ok(date) = v.parse::<Date>() {
        return date.to_zoned(TimeZone::system()).ok();
    }
    let secs: i64 = v.parse().ok()?;
    Some(
        Timestamp::from_second(secs)
            .ok()?
            .to_zoned(TimeZone::system()),
    )
}

/// Renders the elapsed time between `then` and `now` using the largest of the requested
/// `units` that yields a non-zero integer; falls back to the smallest unit (a `0`).
fn relative_between(then: &Zoned, now: &Zoned, units: &[RelativeUnit]) -> Option<String> {
    let mut order: Vec<RelativeUnit> = if units.is_empty() {
        RelativeUnit::all()
    } else {
        units.to_vec()
    };
    // Largest unit first.
    order.sort_by_key(|u| std::cmp::Reverse(u.rank()));
    order.dedup();
    let smallest = *order.last()?;

    let mut chosen = None;
    for &unit in &order {
        let n = elapsed_in(then, now, unit.to_unit())?;
        if n.abs() >= 1 {
            chosen = Some((unit, n));
            break;
        }
    }
    let (unit, n) = match chosen {
        Some(c) => c,
        None => (smallest, elapsed_in(then, now, smallest.to_unit())?),
    };

    let count = n.abs();
    let noun = unit.noun(count);
    Some(if n < 0 {
        format!("in {count} {noun}")
    } else {
        format!("{count} {noun} ago")
    })
}

/// The signed whole number of `unit`s elapsed from `then` to `now` (calendar-aware,
/// truncated toward zero). Positive when `then` is in the past.
fn elapsed_in(then: &Zoned, now: &Zoned, unit: Unit) -> Option<i64> {
    let span = then.until(now).ok()?;
    let total = span.total((unit, then)).ok()?;
    Some(total.trunc() as i64)
}

impl RelativeUnit {
    fn all() -> Vec<Self> {
        use RelativeUnit::{Days, Hours, Minutes, Months, Weeks, Years};
        vec![Minutes, Hours, Days, Weeks, Months, Years]
    }

    /// Ordering rank, smallest unit (minutes) lowest.
    fn rank(self) -> u8 {
        match self {
            Self::Minutes => 0,
            Self::Hours => 1,
            Self::Days => 2,
            Self::Weeks => 3,
            Self::Months => 4,
            Self::Years => 5,
        }
    }

    fn to_unit(self) -> Unit {
        match self {
            Self::Minutes => Unit::Minute,
            Self::Hours => Unit::Hour,
            Self::Days => Unit::Day,
            Self::Weeks => Unit::Week,
            Self::Months => Unit::Month,
            Self::Years => Unit::Year,
        }
    }

    /// The unit's noun, pluralized for any `count` other than 1.
    fn noun(self, count: i64) -> String {
        let base = match self {
            Self::Minutes => "minute",
            Self::Hours => "hour",
            Self::Days => "day",
            Self::Weeks => "week",
            Self::Months => "month",
            Self::Years => "year",
        };
        if count == 1 {
            base.to_string()
        } else {
            format!("{base}s")
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn zoned(y: i16, m: i8, d: i8, h: i8, min: i8) -> Zoned {
        jiff::civil::date(y, m, d)
            .at(h, min, 0, 0)
            .to_zoned(TimeZone::UTC)
            .unwrap()
    }

    #[test]
    fn number_min_max_fraction_digits() {
        let f = Formatter::Number {
            decimal_places: DecimalPlaces::List(vec![0, 1]),
        };
        assert_eq!(f.format("3").as_deref(), Some("3"));
        assert_eq!(f.format("3.0").as_deref(), Some("3"));
        assert_eq!(f.format("3.14").as_deref(), Some("3.1"));
        assert_eq!(f.format("1234567.8").as_deref(), Some("1234567.8"));
        assert_eq!(f.format("-3.0").as_deref(), Some("-3"));
    }

    #[test]
    fn number_keeps_minimum_digits() {
        let f = Formatter::Number {
            decimal_places: DecimalPlaces::List(vec![2, 4]),
        };
        assert_eq!(f.format("3.5").as_deref(), Some("3.50"));
        assert_eq!(f.format("3").as_deref(), Some("3.00"));
    }

    #[test]
    fn number_bad_input_is_none() {
        let f = Formatter::Number {
            decimal_places: DecimalPlaces::default(),
        };
        assert_eq!(f.format(""), None);
        assert_eq!(f.format("N/A"), None);
    }

    #[test]
    fn duration_mm_ss() {
        let f = Formatter::Duration {};
        assert_eq!(f.format("245").as_deref(), Some("4:05"));
        assert_eq!(f.format("245.0").as_deref(), Some("4:05"));
        assert_eq!(f.format("0").as_deref(), Some("0:00"));
        assert_eq!(f.format("66.4").as_deref(), Some("1:06"));
        assert_eq!(f.format("3600").as_deref(), Some("60:00"));
        assert_eq!(f.format("nope"), None);
    }

    #[test]
    fn timestamp_strftime() {
        let f = Formatter::Timestamp {
            format: "%Y-%m-%d".to_string(),
        };
        assert_eq!(
            f.format("2026-06-07 14:30:45").as_deref(),
            Some("2026-06-07")
        );
        assert_eq!(f.format("2026-06-07").as_deref(), Some("2026-06-07"));
        assert_eq!(f.format("not a date"), None);
    }

    #[test]
    fn relative_picks_largest_fitting_unit() {
        let now = zoned(2026, 6, 7, 12, 0);
        let then = zoned(2026, 6, 3, 12, 0); // 4 days earlier
        let out = relative_between(&then, &now, &[]).unwrap();
        assert_eq!(out, "4 days ago");
    }

    #[test]
    fn relative_singular_and_zero_fallback() {
        let now = zoned(2026, 6, 7, 12, 0);
        let one_day = zoned(2026, 6, 6, 12, 0);
        assert_eq!(
            relative_between(&one_day, &now, &[RelativeUnit::Days]).unwrap(),
            "1 day ago"
        );
        let seconds_ago = zoned(2026, 6, 7, 11, 59);
        assert_eq!(
            relative_between(&seconds_ago, &now, &[RelativeUnit::Minutes]).unwrap(),
            "1 minute ago"
        );
        let just_now = zoned(2026, 6, 7, 12, 0);
        assert_eq!(
            relative_between(&just_now, &now, &[RelativeUnit::Minutes]).unwrap(),
            "0 minutes ago"
        );
    }

    #[test]
    fn relative_future_uses_in() {
        let now = zoned(2026, 6, 7, 12, 0);
        let later = zoned(2026, 6, 10, 12, 0); // 3 days later
        assert_eq!(relative_between(&later, &now, &[]).unwrap(), "in 3 days");
    }
}

File: frontend/src/query/format.ts
// Per-column value formatters.
//
// A column's metadata may carry a `formatter` blob selecting a typed transform
// that reshapes each cell's display text. Cell values arrive as plain strings
// (stringified Arrow values), so every formatter parses the string it is given
// and returns `null` when it can't — the caller then falls back to the raw value.
//
// DEVIATION FROM RUST: the reference uses the `jiff` crate for `Timestamp` /
// `RelativeTime`. Per the plan (§3.3) we add no date library and use the built-in
// `Date` plus hand-rolled elapsed math. Consequences:
//   - `Timestamp` supports a reduced `strftime` (the specifiers RadioCrate's
//     queries actually use: %Y %m %d %H %M %S %y %j %% and lowercase am/pm-free),
//     not jiff's full grammar.
//   - `RelativeTime` months/years use average lengths (30.44 / 365.25 days)
//     rather than jiff's calendar-aware spans, so long spans can differ by a day
//     near month boundaries. `whole_app.png` only needs number/duration fidelity.

import type { AnnotationValue } from "querydown-js";
import type { ColumnMetadata } from "./columns";

/** A unit available to the `relativeTime` formatter. */
export type RelativeUnit =
  "minutes" | "hours" | "days" | "weeks" | "months" | "years";

/** The user-facing `decimalPlaces` spec: a plain number, or a one/two-element
 * list of `[min]` / `[min, max]` fraction digits. */
export type DecimalPlaces = number | number[];

/** A typed value formatter, parsed from a column's `formatter` metadata blob.
 * The `type` discriminant selects the variant; each carries its own config. */
export type Formatter =
  | { type: "number"; decimalPlaces: DecimalPlaces }
  | { type: "timestamp"; format: string }
  | { type: "duration" }
  | { type: "relativeTime"; units: RelativeUnit[] };

const RELATIVE_UNITS: RelativeUnit[] = [
  "minutes",
  "hours",
  "days",
  "weeks",
  "months",
  "years",
];

/** Parses a `formatter` metadata blob into a {@link Formatter}, or `undefined`
 * when it's malformed or of an unknown type. Mirrors serde's `tag = "type",
 * rename_all = "camelCase"`: `Number.decimalPlaces` and `RelativeTime.units`
 * default when absent; `Timestamp.format` is required. */
export function parseFormatter(
  blob: AnnotationValue | undefined,
): Formatter | undefined {
  if (blob == null || typeof blob !== "object" || Array.isArray(blob)) {
    return undefined;
  }
  const o = blob as Record<string, AnnotationValue>;
  switch (o.type) {
    case "number":
      return {
        type: "number",
        decimalPlaces: parseDecimalPlaces(o.decimalPlaces),
      };
    case "duration":
      return { type: "duration" };
    case "timestamp":
      if (typeof o.format !== "string") return undefined;
      return { type: "timestamp", format: o.format };
    case "relativeTime":
      return { type: "relativeTime", units: parseUnits(o.units) };
    default:
      return undefined;
  }
}

function parseDecimalPlaces(v: AnnotationValue | undefined): DecimalPlaces {
  if (typeof v === "number") return v;
  if (Array.isArray(v))
    return v.filter((x): x is number => typeof x === "number");
  return [];
}

function parseUnits(v: AnnotationValue | undefined): RelativeUnit[] {
  if (!Array.isArray(v)) return [];
  return v.filter((x): x is RelativeUnit =>
    RELATIVE_UNITS.includes(x as RelativeUnit),
  );
}

/** Formats `value` for display, or returns `null` when it can't be parsed (the
 * caller then renders the raw value unchanged). */
export function formatValue(f: Formatter, value: string): string | null {
  switch (f.type) {
    case "number": {
      const [min, max] = resolveDecimals(f.decimalPlaces);
      return formatNumber(value, min, max);
    }
    case "duration":
      return formatDuration(value);
    case "timestamp":
      return formatTimestamp(value, f.format);
    case "relativeTime":
      return formatRelative(value, f.units, Date.now());
  }
}

/** Applies the column's formatter and prefix/suffix to one raw cell value. */
export function displayText(meta: ColumnMetadata, value: string): string {
  const formatted = meta.formatter
    ? (formatValue(meta.formatter, value) ?? value)
    : value;
  if (meta.prefix === "" && meta.suffix === "") return formatted;
  return `${meta.prefix}${formatted}${meta.suffix}`;
}

/** Resolves a {@link DecimalPlaces} spec into a `[min, max]` pair, swapping if
 * min > max. */
function resolveDecimals(spec: DecimalPlaces): [number, number] {
  let min: number;
  let max: number;
  if (typeof spec === "number") {
    min = spec;
    max = spec;
  } else if (spec.length === 0) {
    min = 0;
    max = 0;
  } else if (spec.length === 1) {
    min = spec[0];
    max = spec[0];
  } else {
    min = spec[0];
    max = spec[1];
  }
  if (min > max) [min, max] = [max, min];
  return [min, max];
}

/** Formats a number to at most `max` and at least `min` fraction digits, trimming
 * trailing zeros in between. No thousands grouping. */
function formatNumber(value: string, min: number, max: number): string | null {
  const trimmed = value.trim();
  // `Number("")` is 0, not NaN — guard the empty string like Rust's parse.
  if (trimmed === "") return null;
  const n = Number(trimmed);
  if (Number.isNaN(n)) return null;
  let s = n.toFixed(max);
  // `s` now has exactly `max` fraction digits. Trim trailing zeros down to `min`.
  if (max > min) {
    const dot = s.indexOf(".");
    if (dot !== -1) {
      const minLen = min === 0 ? dot : dot + 1 + min;
      while (s.length > minLen && s.endsWith("0")) s = s.slice(0, -1);
      if (s.endsWith(".")) s = s.slice(0, -1);
    }
  }
  return s;
}

/** Formats a number of seconds as `M:SS` (minutes unpadded, seconds zero-padded),
 * rounded to the nearest second. */
function formatDuration(value: string): string | null {
  const trimmed = value.trim();
  if (trimmed === "") return null;
  const secs = Number(trimmed);
  if (Number.isNaN(secs)) return null;
  const total = Math.round(secs);
  const sign = total < 0 ? "-" : "";
  const abs = Math.abs(total);
  const minutes = Math.floor(abs / 60);
  const seconds = abs % 60;
  return `${sign}${minutes}:${String(seconds).padStart(2, "0")}`;
}

/** Civil datetime fields parsed from a cell value. */
interface Civil {
  year: number;
  month: number;
  day: number;
  hour: number;
  minute: number;
  second: number;
}

const DATETIME_RE = /^(\d{4})-(\d{2})-(\d{2})[ T](\d{2}):(\d{2}):(\d{2})/;
const DATE_RE = /^(\d{4})-(\d{2})-(\d{2})$/;

/** Parses a cell value into civil datetime fields. Tries a civil datetime, then a
 * date, then integer epoch seconds (converted in local time). Mirrors the
 * fallback chain of `format_timestamp` / `parse_zoned`. */
function parseCivil(value: string): Civil | null {
  const v = value.trim();
  const dt = DATETIME_RE.exec(v);
  if (dt) {
    return {
      year: +dt[1],
      month: +dt[2],
      day: +dt[3],
      hour: +dt[4],
      minute: +dt[5],
      second: +dt[6],
    };
  }
  const d = DATE_RE.exec(v);
  if (d) {
    return {
      year: +d[1],
      month: +d[2],
      day: +d[3],
      hour: 0,
      minute: 0,
      second: 0,
    };
  }
  if (/^-?\d+$/.test(v)) {
    const date = new Date(Number(v) * 1000);
    if (!Number.isNaN(date.getTime())) {
      return {
        year: date.getFullYear(),
        month: date.getMonth() + 1,
        day: date.getDate(),
        hour: date.getHours(),
        minute: date.getMinutes(),
        second: date.getSeconds(),
      };
    }
  }
  return null;
}

/** Parses a cell value into epoch milliseconds (local time for civil values), for
 * relative-time math. */
function parseEpochMs(value: string): number | null {
  const c = parseCivil(value);
  if (!c) return null;
  const date = new Date(c.year, c.month - 1, c.day, c.hour, c.minute, c.second);
  const ms = date.getTime();
  return Number.isNaN(ms) ? null : ms;
}

const pad2 = (n: number) => String(n).padStart(2, "0");

/** A reduced `strftime`: supports the specifiers RadioCrate's queries use. */
function strftime(c: Civil, format: string): string {
  return format.replace(/%(.)/g, (_m, spec: string) => {
    switch (spec) {
      case "Y":
        return String(c.year);
      case "y":
        return pad2(c.year % 100);
      case "m":
        return pad2(c.month);
      case "d":
        return pad2(c.day);
      case "H":
        return pad2(c.hour);
      case "M":
        return pad2(c.minute);
      case "S":
        return pad2(c.second);
      case "%":
        return "%";
      default:
        return `%${spec}`;
    }
  });
}

function formatTimestamp(value: string, format: string): string | null {
  const c = parseCivil(value);
  if (!c) return null;
  return strftime(c, format);
}

const MS: Record<RelativeUnit, number> = {
  minutes: 60_000,
  hours: 3_600_000,
  days: 86_400_000,
  weeks: 604_800_000,
  months: 2_629_800_000, // 30.44 days (average — see file-level deviation note)
  years: 31_557_600_000, // 365.25 days
};

const RANK: Record<RelativeUnit, number> = {
  minutes: 0,
  hours: 1,
  days: 2,
  weeks: 3,
  months: 4,
  years: 5,
};

const NOUN: Record<RelativeUnit, string> = {
  minutes: "minute",
  hours: "hour",
  days: "day",
  weeks: "week",
  months: "month",
  years: "year",
};

function formatRelative(
  value: string,
  units: RelativeUnit[],
  nowMs: number,
): string | null {
  const thenMs = parseEpochMs(value);
  if (thenMs === null) return null;
  return relativeBetween(thenMs, nowMs, units);
}

/** The signed whole number of `unit`s elapsed from `then` to `now` (truncated
 * toward zero). Positive when `then` is in the past. */
function elapsedIn(thenMs: number, nowMs: number, unit: RelativeUnit): number {
  return Math.trunc((nowMs - thenMs) / MS[unit]);
}

/** Renders the elapsed time between `then` and `now` using the largest of the
 * requested `units` that yields a non-zero integer; falls back to the smallest
 * unit (a `0`). Exported for unit tests (mirrors `relative_between`). */
export function relativeBetween(
  thenMs: number,
  nowMs: number,
  units: RelativeUnit[],
): string | null {
  const source = units.length === 0 ? RELATIVE_UNITS : units;
  // Largest unit first, deduped.
  const order = [...new Set(source)].sort((a, b) => RANK[b] - RANK[a]);
  const smallest = order[order.length - 1];
  if (smallest === undefined) return null;

  let chosen: [RelativeUnit, number] | undefined;
  for (const unit of order) {
    const n = elapsedIn(thenMs, nowMs, unit);
    if (Math.abs(n) >= 1) {
      chosen = [unit, n];
      break;
    }
  }
  const [unit, n] = chosen ?? [smallest, elapsedIn(thenMs, nowMs, smallest)];

  const count = Math.abs(n);
  const noun = count === 1 ? NOUN[unit] : `${NOUN[unit]}s`;
  return n < 0 ? `in ${count} ${noun}` : `${count} ${noun} ago`;
}
